### PR TITLE
Performance

### PR DESCRIPTION
**Pull request by photon1503**
Performance impact on my machine when running the plugin is about 10% CPU, so I started to analyze.

When checking the current Plugin SDK Demo, we can see that AddProperty / SetProperty has been replaced by AttachDelegate.
See also the _SimHub plugin 101_ thread at https://www.racedepartment.com/threads/simhub-plugin-s-for-output-to-midi-and-vjoy.210079/

So this new way of using the properties is way more performant then the "old" way with Add/Set.

I started to change the Dashboard.cs and some other entries and it looks fine.

### CODE REVIEW

**Comment by andreasdahl1987** [comment]
Good catch! How much CPU is being used when runing SimHub without my plugin? The AttachDelegate() is something I've been aware of but never took the time to convert to :see_no_evil: Excited to see if it has a major performance impact! Other than that I'm suspecting the main culprit is the delta calculations, which uses a ton of for loops to update up to 4000 variables, and also has to be calculated on every cycle. I've written the algorithm in a way that only relevant variables will be calculated for and updated, but I think it still uses a lot of processing power. It is an essential part of the plugin though.

**Comment by photon1503** [comment]
without the plugin running I have a utilization of 1,8% CPU.
With the latest version the counter goes up to 3% now.

**Comment by andreasdahl1987** [comment]
Oh that's a huge improvement then, as long as it still works as intended 😅

**Comment by photon1503** [comment]
Yeah, testing is needed. Now trying to fix a bug which I just found

**Comment by photon1503** [comment]
Looks now ok from my side, please review and test when you have time

**Comment by andreasdahl1987** [comment]
Will do @photon1503 !

File: Sections/DDC.cs [hunk 1]
```diff
@@ -101,15 +102,15 @@ public DDC(DahlDesign dahlDesign)
...
-            Base.AddProp("SW1startLED", Base.Settings.SW1startLED);
-            Base.AddProp("DDUEnabled",  Base.Settings.DDUEnabled);
+            Base.AttachDelegate("DDUstartLED", () => Base.Settings.DDUstartLED);
+            Base.AttachDelegate("SW1startLED", () => Base.Settings.SW1startLED);
+            Base.AttachDelegate("DDUEnabled", () => Base.Settings.DDUEnabled);
         }
 
         public void DataUpdate()
         {
-            Base.SetProp("DDCDDSEnabled",  false);
```
Comment (andreasdahl1987) [inline Sections/DDC.cs:111]: So I think this breaks the DDS integration with the plugin menu. It is always false and I cant control the plugin menu with a DDS (Dahl Design Switch, a special kind of firmware for an encoder available in DDC)

File: iRacing/iRacing.cs [hunk 2]
```diff
@@ -1915,14 +1858,16 @@ public void DataUpdate()
...
                 {
-                    iratings.Add(0);
+                    if (Base.gameData.NewData.Opponents[i].CarClass == myClass)
+                    {
+                        iratings.Add(Base.gameData.NewData.Opponents[i].IRacing_IRating);
+                    }
+                    else
+                    {
+                        iratings.Add(0);
+                    }
```
Comment (andreasdahl1987) [inline iRacing/iRacing.cs:1870]: So this looks like it fixed the iRating issue, hadn't noticed it was broken. you know when this happened? 😅

File: iRacing/iRacing.cs [hunk 3]
```diff
@@ -666,108 +681,36 @@ public Data(DahlDesign dahlDesign)
...
+                Base.AddProp($"Lap{propIndex}Sector1Delta", 0);
+                Base.AddProp($"Lap{propIndex}Sector2Delta", 0);
+                Base.AddProp($"Lap{propIndex}Sector3Delta", 0);
+                Base.AttachDelegate($"Lap{propIndex}Sector1Status", () => sector1StatusList[i]);
+                Base.AttachDelegate($"Lap{propIndex}Sector2Status", () => sector2StatusList[i]);
+                Base.AttachDelegate($"Lap{propIndex}Sector3Status", () => sector3StatusList[i]);
+                Base.AttachDelegate($"Lap{propIndex}FuelTargetDelta", () => fuelTargetDeltas[i]);
+            }
+
+            Base.AttachDelegate("LapRecord", () => lapRecord);
```
Comment (andreasdahl1987) [inline iRacing/iRacing.cs:713]: Can't decifer it completely from the changes, but lap recod delta is broken.

File: iRacing/iRacing.cs [hunk 4]
```diff
@@ -608,50 +623,50 @@ public Data(DahlDesign dahlDesign)
...
 
-            Base.AddProp("PitBoxPosition", 1);
-            Base.AddProp("PitBoxApproach", false);
+            Base.AttachDelegate("PitBoxPosition", () => pitBox);  //TODO: check, was 1 before
+            Base.AttachDelegate("PitBoxApproach", () => boxApproach);
             Base.AddProp("PitEntry", false);
             Base.AddProp("PitSpeeding", false);
 
-            Base.AddProp("SessionBestLap", new TimeSpan(0));
+            Base.AttachDelegate("SessionBestLap", () => lapRecord);
```
Comment (andreasdahl1987) [inline iRacing/iRacing.cs:663]: When setting new session best, it is set to lap record

**Review by andreasdahl1987 — COMMENTED** [review]
Done some testing, here are some comments

**Comment by andreasdahl1987** [comment]
Will have to do some more reviewing tomorrow 👍

File: Sections/DDC.cs (hunk 1 above)
Comment (photon1503) [reply Sections/DDC.cs:111]: I compared this to some older commit.

The property DDCDDSEnabled was implicitly set to false on each iteration and later set to the value of Settings.DDSEnabled.

Is there any reason for this? I will now simply remove the implicit setting of the value DDSEnabled, it should be sufficiant that we have the binding from the value to the property using the delegate and also the connection from the settings page (XAML) to the property.

DDC.cs:
`Base.AttachDelegate("DDCDDSEnabled",  () => Base.Settings.DDSEnabled);`

will remove this (and also for DDCclutchEnabled, same issue here)
`Base.Settings.DDSEnabled = false;`

SettingsControl.xaml:
` <TextBlock Grid.Row="2" Text="DDS Enabled" />
                                <styles:SHToggleButton Grid.Row="2" Grid.Column="1" HorizontalAlignment="Left" VerticalAlignment="Top" Height="20" IsChecked="{Binding DDSEnabled}"/>
                               `

File: iRacing/iRacing.cs (hunk 2 above)
Comment (photon1503) [reply iRacing/iRacing.cs:1870]: this must have happend on Nov, 10th when I made the refactoring. I messed up the loop logic

File: iRacing/iRacing.cs (hunk 4 above)
Comment (photon1503) [reply iRacing/iRacing.cs]: fixed

File: iRacing/iRacing.cs (hunk 3 above)
Comment (photon1503) [reply iRacing/iRacing.cs:713]: lapRecord is being updated by LapRecords class, using a reference. Maybe this doesn't work with delegates. So I created a dedicated variable 'proplapRecords' to assign

**Comment by photon1503** [comment]
@andreasdahl1987 : Please review the PR again, tried to fix all issues.

**Comment by andreasdahl1987** [comment]
Great! I'll get some time to do tests in the evening

File: iRacing/iRacing.cs (hunk 3 above)
Comment (photon1503) [reply iRacing/iRacing.cs]: Now I have seen the issue, the properties Lap<nn>Sector<n>Delta are not updated correctly :-(

Comment (andreasdahl1987) [reply iRacing/iRacing.cs]: Ok, maybe leave the lap timing system as it was?

File: Sections/DDC.cs (hunk 1 above)
Comment (andreasdahl1987) [reply Sections/DDC.cs:111]: I can't really say why I did that. It was one of the first settings I set up with a toggle in the menu. I might have had some misinterpretation of how the system works

File: iRacing/iRacing.cs (hunk 3 above)
Comment (andreasdahl1987) [reply iRacing/iRacing.cs:713]: image
 Hmm, testing now it seems like all is in order.

Comment (photon1503) [reply iRacing/iRacing.cs:713]: Also now check the lap records. Modifed the CSV file and rerun a session. The lap was dedected as new lap record and the CSV was update.

image

**Comment by photon1503** [comment]
This PR now also includes a new way to increase the counter. Currently it is increased by 1 every time the plugin is called by the host (SimHub) and reset when > 59.

Instead, the execution of the whole update procedure now limits to 60 FPS using tick calculation.
We should also think to make this update frequency as an option. I.e. I'm using a Vocore display which has only 24Hz.

```
            //FRAME COUNTER FOR CPU SAVING
            long nowTicks = DateTime.Now.Ticks;
            UpdateAt60Fps = nowTicks - LastRan60Fps >= TicksFor60Fps;
            if (UpdateAt60Fps)
            {
                LastRan60Fps = DateTime.Now.Ticks;
                counter++;
                //Resetting counter
                if (counter > 59)
                {
                    counter = 0;
                }
            } else
            {
                return;
            }
```

File: Sections/DDC.cs (hunk 1 above)
Comment (photon1503) [reply Sections/DDC.cs:111]: @andreasdahl1987 just let me know if it works now, then we can resolve this point

**Comment by photon1503** [comment]
For debugging performance, I added some performance counters.

image

Here we can see the spent time for each "counter" iteration and also the sum of them. 

I included them in this PR, but they will only be calculated in Debug mode.

**Comment by andreasdahl1987** [comment]
This way of setting the frequency...the dataflow and calculations is still 60 Hz, right? Just the updating of the properties that we can slow down? Cus the rate of data refresh in many cases is the resolution of tracked variables - such as delta times, sector times, emulation of traction control and many more.

**Comment by andreasdahl1987** [comment]
Btw the slower I am at merging your PRs the bigger they get :D Your contribution to this project is really something

**Comment by andreasdahl1987** [comment]
> For debugging performance, I added some performance counters.
> 
> image
> 
> Here we can see the spent time for each "counter" iteration and also the sum of them.
> 
> I included them in this PR, but they will only be calculated in Debug mode.

What exactly is this telling me? Which "counter" has more headroom? The sum, can I compare it to anything, or is it soft of my own reference for adding new features as checking how it affects performance?

**Comment by andreasdahl1987** [comment]
Are you planning on adding more, I think I'm ready to merge this

**Comment by photon1503** [comment]
> Are you planning on adding more? I think I'm ready to merge this

No, let's merge it. It's already a big PR and should be closed

**Comment by photon1503** [comment]
> > For debugging performance, I added some performance counters.
> > image
> > Here we can see the spent time for each "counter" iteration and also the sum of them.
> > I included them in this PR, but they will only be calculated in Debug mode.
> 
> What exactly is this telling me? Which "counter" has more headroom? The sum, can I compare it to anything, or is it soft of my own reference for adding new features as checking how it affects performance?

The counters are nice for debugging and will show us how much time is spent for each of the calculations. Most of them are finished in 1-2ms, some are taking longer.

The sum is artificial, since not all functions are called at every iteration. Maybe useless...

**Comment by photon1503** [comment]
> This way of setting the frequency...the dataflow and calculations is still 60 Hz, right? Just the updating of the properties that we can slow down? Cus the rate of data refresh in many cases is the resolution of tracked variables - such as delta times, sector times, emulation of traction control and many more.

Yes, the frequency is still 60 Hz. Just skipping those calls which would be faster than 60Hz.

**Comment by andreasdahl1987** [comment]
There is a big loss in resolution of the delta change calculations. Related to refreshrate? Last time I tested before the performance counters were added, it was fine. Now a recorded lap record will have anything from -0.05 to +0.05 seconds in delta to itself. Used to be a steady 0.00

**Comment by photon1503** [comment]
@andreasdahl1987 ok I see. Than let's remove the check, will do a PR later today. Obviously we must not skip any frame since the plugin is called at a 60Hz rate already by SimHub

**Pull request merged**